token/program/src/state.rs

```diff
@@ -287,6 +287,56 @@ fn unpack_coption_u64(src: &[u8; 12]) -> Result<COption<u64>, ProgramError> {
     }
 }
 
+const SPL_TOKEN_ACCOUNT_MINT_OFFSET: usize = 0;
+const SPL_TOKEN_ACCOUNT_OWNER_OFFSET: usize = 32;
+
+/// A trait for token Account structs to enable efficiently unpacking various fields
+/// without unpacking the complete state.
+pub trait GenericTokenAccount {
+    /// Check if the account data is a valid token account
+    fn valid_account_data(account_data: &[u8]) -> bool;
+
+    /// Call after account length has already been verified to unpack the account owner
+    fn unpack_account_owner_unchecked(account_data: &[u8]) -> &Pubkey {
+        Self::unpack_pubkey_unchecked(account_data, SPL_TOKEN_ACCOUNT_OWNER_OFFSET)
+    }
+
+    /// Call after account length has already been verified to unpack the account mint
+    fn unpack_account_mint_unchecked(account_data: &[u8]) -> &Pubkey {
+        Self::unpack_pubkey_unchecked(account_data, SPL_TOKEN_ACCOUNT_MINT_OFFSET)
+    }
+
+    /// Call after account length has already been verified to unpack a Pubkey at
+    /// the specified offset. Panics if `account_data.len()` is less than `PUBKEY_BYTES`
+    fn unpack_pubkey_unchecked(account_data: &[u8], offset: usize) -> &Pubkey {
+        bytemuck::from_bytes(&account_data[offset..offset + PUBKEY_BYTES])
+    }
+
+    /// Unpacks an account's owner from opaque account data.
+    fn unpack_account_owner(account_data: &[u8]) -> Option<&Pubkey> {
+        if Self::valid_account_data(account_data) {
+            Some(Self::unpack_account_owner_unchecked(account_data))
+        } else {
+            None
+        }
+    }
+
+    /// Unpacks an account's mint from opaque account data.
+    fn unpack_account_mint(account_data: &[u8]) -> Option<&Pubkey> {
+        if Self::valid_account_data(account_data) {
+            Some(Self::unpack_account_mint_unchecked(account_data))
+        } else {
+            None
+        }
+    }
+}
+
+impl GenericTokenAccount for Account {
+    fn valid_account_data(account_data: &[u8]) -> bool {
+        account_data.len() == Account::LEN
```

Comment:
Sorry for the late review here, but to be safe, this could also check that `is_initialized` is properly set, as in https://github.com/solana-labs/solana-program-library/blob/bbbf250a67777a3dea8717648bf1b78d72fa1a8c/token/program-2022/src/extension/mod.rs#L217-L224

Reply:
Thanks for the suggestion -- already merged code but I will have another PR to address it.

Reply:
Yeah I noticed, wanted to make you aware either way!

Reply:
Hi Jon, the code you pointed to is in the program-2022 module, is that relevant for the regular SPL token? .i.e the ACCOUNT_INITIALIZED_INDEX position and the assumption of it need to be initialized to be valid?

Secondly -- if I understand correctly, the account extension can have more data then Account::Len -- is the following logic still good?

Reply:
> the code you pointed to is in the program-2022 module, is that relevant for the regular SPL token? .i.e the ACCOUNT_INITIALIZED_INDEX position and the assumption of it need to be initialized to be valid?

Yes, it is still relevant for regular SPL token.

> Secondly -- if I understand correctly, the account extension can have more data then Account::Len -- is the following logic still good?

For normal SPL token, it's perfect, but correct, token-2022 can have a longer length, so that would not work.  You'd have to do some logic like you did here:

```
    fn valid_account_data(account_data: &[u8]) -> bool {
        spl_token::state::Account::valid_account_data(account_data)
            || ACCOUNTTYPE_ACCOUNT == *account_data.get(Account::LEN).unwrap_or(&0)
    }
```